pkg/controllers/ratelimitpolicy/ratelimitpolicy_controller.go

```diff
@@ -0,0 +1,206 @@
+*/
+
+package ratelimitpolicy
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+	gatewayv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+
+	kuadrantapi "github.com/kuadrant/kuadrant-operator/api/v1beta1"
+	kuadrantcommon "github.com/kuadrant/kuadrant-operator/pkg/common"
+
+	"github.com/Kuadrant/multi-cluster-traffic-controller/pkg/_internal/clusterSecret"
+	syncutils "github.com/Kuadrant/multi-cluster-traffic-controller/pkg/_internal/sync"
+	"github.com/Kuadrant/multi-cluster-traffic-controller/pkg/syncer"
+)
+
+const (
+	RateLimitPolicyFinalizer = "kuadrant.io/rate-limit-policy"
+)
+
+// RateLimitPolicyReconciler reconciles a RateLimitPolicy object
+type RateLimitPolicyReconciler struct {
+	client.Client
+	Scheme         *runtime.Scheme
+	ClusterSecrets *clusterSecret.Service
+}
+
+//+kubebuilder:rbac:groups=kuadrant.io,resources=ratelimitpolicies,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=kuadrant.io,resources=ratelimitpolicies/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=kuadrant.io,resources=ratelimitpolicies/finalizers,verbs=update
+
+func (r *RateLimitPolicyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	_ = log.FromContext(ctx)
+
+	previous := &kuadrantapi.RateLimitPolicy{}
+	err := r.Client.Get(ctx, client.ObjectKey{Namespace: req.Namespace, Name: req.Name}, previous)
+	if err != nil {
+		if err := client.IgnoreNotFound(err); err == nil {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+	rateLimitPolicy := previous.DeepCopy()
+
+	log.Log.V(3).Info("RateLimitPolicyReconciler Reconcile", "rateLimitPolicy", rateLimitPolicy)
+
+	if rateLimitPolicy.DeletionTimestamp != nil && !rateLimitPolicy.DeletionTimestamp.IsZero() {
+		controllerutil.RemoveFinalizer(rateLimitPolicy, RateLimitPolicyFinalizer)
+
+		err = r.Update(ctx, rateLimitPolicy)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	targetGateway, err := r.getTargetGateway(ctx, rateLimitPolicy)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	err = controllerutil.SetControllerReference(targetGateway, rateLimitPolicy, r.Scheme)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if !controllerutil.ContainsFinalizer(rateLimitPolicy, RateLimitPolicyFinalizer) {
+		log.Log.Info("RateLimitPolicyFinalizer Reconcile: Added finalizer and set gateway owner", "rateLimitPolicy", rateLimitPolicy.Name)
+
+		controllerutil.AddFinalizer(rateLimitPolicy, RateLimitPolicyFinalizer)
+
+		err = r.Update(ctx, rateLimitPolicy)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
```

Comment:
should we add a return after this if as I think the update will trigger a new reconcile and currently this reconcile would continue?